<h1><center>Laboratorio 4: El Pandas no muerde (act. II) 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Pablo Badilla
- Auxiliar: Ignacio Meza D.
- Ayudante: Constanza Peña

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Camila J. Pulgar Fernández
- Nombre de alumno 2: Francisca Solis G. 

### **Link de repositorio de GitHub:** `https://github.com/cpulgarf/Laboratorios`

# Temas a tratar

- Manejo de datos tabulares usando `pandas`. En esta segunda parte se incluye adicionalmente agregaciones, concatenaciones, merge y trabajo con strings.
- Visualizaciones interactivas de los datos con `plotly`.

### Reglas:

- Fecha de entrega: 8 de octubre (atrasos hasta el 11 de octubre, 1 punto de descuento por día)
- **Grupos de máximo 2 personas**
- **Ausentes** deberán realizar la actividad solos. 
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquer matrial del curso que estimen conveniente.


### Objetivos principales del laboratorio

- Aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas`.
- Utilizar `plotly` para obtener información gráfica del dataset.
- Aplicar el **Análisis Exploratorio de Datos** a un caso en particular.


> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

# Importamos librerias utiles 😸

In [2]:
# Libreria Core del lab.
import numpy as np
import pandas as pd 

from IPython.display import display
#Libreria para plotear
import plotly.graph_objects as go

#Libreria para visualizar
#!pip install --upgrade plotly
import plotly.figure_factory as ff
import plotly.express as px

# 1. Rendimiento en Estudiantes 📚


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

Para este laboratorio deberán continuar el Análisis Exploratorio de datos sobre el conjunto ```students_grades.csv```, el cual contiene una caracterización sobre el rendimiento y otros atributos de cada alumno de la *Universidad de la Cachaña* .

In [ ]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = '/content/drive/MyDrive/Lab4'
except: 
    print('Ignorando conexión drive-colab')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# 1. Carga de Datos [0.5 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente). 

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: ´students_grades_1.json´ y ´students_grades_2.json´. 


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos. No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas, recuerde que anteriormente ya se encargo de este tema.

In [184]:
from google.colab import files
uploaded = files.upload()

Saving other_grades.csv to other_grades.csv
Saving students_grades.json to students_grades.json
Saving students_grades_1.json to students_grades_1 (1).json
Saving students_grades_2.json to students_grades_2 (1).json


In [5]:
df_grades_1 = pd.read_json("students_grades_1.json")
df_grades_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   names                        400 non-null    object 
 1   gender                       400 non-null    object 
 2   race/ethnicity               400 non-null    object 
 3   parental level of education  400 non-null    object 
 4   lunch                        400 non-null    object 
 5   test preparation course      400 non-null    object 
 6   math score                   400 non-null    float64
 7   reading score                400 non-null    float64
 8   writing score                400 non-null    float64
dtypes: float64(3), object(6)
memory usage: 28.2+ KB


In [6]:
df_grades_2 = pd.read_json("students_grades_2.json")
df_grades_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 475 entries, 0 to 474
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   names                        475 non-null    object 
 1   gender                       475 non-null    object 
 2   race/ethnicity               475 non-null    object 
 3   parental level of education  475 non-null    object 
 4   lunch                        475 non-null    object 
 5   test preparation course      475 non-null    object 
 6   math score                   475 non-null    float64
 7   reading score                475 non-null    float64
 8   writing score                475 non-null    float64
dtypes: float64(3), object(6)
memory usage: 33.5+ KB


In [95]:
df_grades = pd.concat([df_grades_1,df_grades_2],axis = 0)

In [10]:
df_grades

names  gender  ... reading score writing score
0       Rita Courtney  female  ...          3.76          3.76
1    Charles Linstrom    male  ...          5.68          5.86
2         Brian Young    male  ...          4.96          4.78
3      Howard Jimenez    male  ...          5.50          5.56
4        Wayne Wilson    male  ...          6.16          6.22
..                ...     ...  ...           ...           ...
470     Richard Young    male  ...          5.50          5.26
471     Wanda Russell  female  ...          3.64          3.16
472    Marina Zeigler  female  ...          5.44          5.86
473     Laurie Carter  female  ...          4.66          4.72
474      Amanda Perez  female  ...          5.80          5.56

[875 rows x 9 columns]

Resultado esperado:

In [ ]:
#df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
470     Richard Young    male        group D                 high school   
471     Wanda Russell  female        group B                 high school   
472    Marina Zeigler  female        group C           bachelor's degree   
473     Laurie Carter  female        group B            some high school   
474      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
470      standard                    none        5.14           5.50   
471  free/reduced               completed        2.38           3.64   
472  free/reduced               completed        4.96           5.44   
473      standard               completed        4.24           4.66   
474      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
470           5.26  
471           3.16  
472           5.86  
473           4.72  
474           5.56  

[875 rows x 9 columns]

## Análisis de Las Notas v2 [1.5 Puntos]

Preocupado por la dificultad que representan los boxplot, el rector le solicita buscar alternativas de visualización la distribución de las notas. 

Para esto, muestre el [boxplot](https://plotly.com/python/box-plots/) mostrado la clase anterior, utilice un [displot](https://plotly.com/python/distplot/#distplot-with-pandas) como también un [histograma con un gráfico marginal de caja](https://plotly.com/python/histograms/#visualizing-the-distribution) como también un [histograma con el ramo como faceta de fila](https://plotly.com/python/facet-plots/#histogram-facet-grids).

Luego, responda las siguientes pregunta: 

> 1. ¿Existe una diferencia notable entre las notas?

> 2. ¿Cuál de los gráficos mostrados cree que es adecuado para mostrarle al rector? ¿Y a los padres? ¿Y a un centro de estudios educativos? ¿Por qué?. Base sus respuestas en lo visto en la clase de visualizaciones como también en lo que usted y su equipo consideren correcto.

> Hint: Para elaborar el histograma, puede que le sea de utilidad hacer un `melt` del DataFrame, dejando como variables los ramos y valores las notas. Por otra parte, visiten la documentación para generar los gráficos.


**Gráfico de Caja:**

In [96]:
df_grades["math score"] = pd.to_numeric(df_grades["math score"])
df_grades["reading score"] = pd.to_numeric(df_grades["reading score"])
df_grades["writing score"] = pd.to_numeric(df_grades["writing score"])

In [12]:
fig = go.Figure()
fig.add_trace(go.Box(y=df_grades.iloc[:,6], name='math score'))
fig.add_trace(go.Box(y=df_grades.iloc[:,7], name = 'reading score',
                marker_color = 'indianred'))
fig.add_trace(go.Box(y=df_grades.iloc[:,8], name = 'writing score',
                marker_color = 'lightseagreen'))
fig.show()

**Distplot:**

In [13]:
# use displot figurefactory
df = df_grades.loc[:,['math score','reading score','writing score']]
fig = ff.create_distplot([df[c] for c in df.columns], df.columns, show_hist = False)
fig.show()

**Histograma con Boxplots:**

In [14]:
df = df_grades.loc[:,['math score','reading score','writing score']]
df_m = pd.melt(df)
fig = px.histogram(df_m , x = 'value', y = 'variable', color = 'variable',  marginal="box", barmode = 'group', nbins = 15)  
fig.show()

**Histograma con Faceta:**

In [15]:
df = df_grades.loc[:,['math score','reading score','writing score']]
df_m = pd.melt(df)
fig = px.histogram(df_m , x = 'value', y = 'variable', color = 'variable', facet_row = 'variable', nbins=20)
fig.show()

**Resultados Esperados:**

**Justifique:**

```

```

## Análisis por Nivel Educacional Etnia de los Padres [1.5 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos. 
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.

### Tabla de Resumen

Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar `map()`) de la variable `parental level of education`  según la siguiente conversión: 

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas. 
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'. 


Utilizar la tabla de resultados esperados como guía para desarrollar este punto.

### 1.3.4 Información por Categorías [1 punto]

In [97]:
def alumnos_promedio(dataframe_in):
  dataframe_in["GPA"] = pd.to_numeric(round(dataframe_in[["math score","reading score","writing score"]].mean(axis=1), 2))
  return dataframe_in

In [98]:
df_grades["simple parental level of education"] = df_grades['parental level of education'].map({"some high school": "school",
                                                                                               "some college" : "school",
                                                                                               "high school"  : "school",
                                                                                               "bachelor's degree" : "college",
                                                                                               "associate's degree" : "college",
                                                                                               "master's degree" : "postgraduate"})

In [150]:
df_grades

names  gender  ... writing score simple parental level of education
0       Rita Courtney  female  ...          3.76                             school
1    Charles Linstrom    male  ...          5.86                            college
2         Brian Young    male  ...          4.78                             school
3      Howard Jimenez    male  ...          5.56                             school
4        Wayne Wilson    male  ...          6.22                             school
..                ...     ...  ...           ...                                ...
470     Richard Young    male  ...          5.26                             school
471     Wanda Russell  female  ...          3.16                             school
472    Marina Zeigler  female  ...          5.86                            college
473     Laurie Carter  female  ...          4.72                             school
474      Amanda Perez  female  ...          5.56                             school

[875 rows x 10 columns]

In [126]:
df = round(df_grades.groupby(['race/ethnicity','simple parental level of education']).mean(),2)

In [127]:
df = alumnos_promedio(df)

In [128]:
x = round(df_grades.groupby(['race/ethnicity','simple parental level of education'])['names'].count(),2)

In [129]:
df['count'] = x

In [130]:
df

math score  ...  count
race/ethnicity simple parental level of education              ...       
group A        college                                   4.73  ...     24
               postgraduate                              4.69  ...      2
               school                                    4.57  ...     51
group B        college                                   5.07  ...     54
               postgraduate                              4.91  ...      5
               school                                    4.69  ...    107
group C        college                                   5.02  ...    102
               postgraduate                              4.92  ...     15
               school                                    4.76  ...    155
group D        college                                   5.11  ...     70
               postgraduate                              5.22  ...     20
               school                                    5.02  ...    149
group E        college                                   5.54  ...     52
               postgraduate                              5.54  ...      6
               school                                    5.40  ...     63

[15 rows x 5 columns]

In [136]:
def percentage(dataframe_in):
  a= (round((dataframe_in['count'] / dataframe_in['count'].sum()) * 100, 2)).astype(str)
  dataframe_in['percentage'] =  a + "%"
  return dataframe_in


In [139]:
df2 = percentage(df)

In [156]:
df2.columns.values

array(['math score', 'reading score', 'writing score', 'GPA', 'count',
       'percentage'], dtype=object)

In [152]:
df_grades.head(4)

names  gender  ... writing score simple parental level of education
0     Rita Courtney  female  ...          3.76                             school
1  Charles Linstrom    male  ...          5.86                            college
2       Brian Young    male  ...          4.78                             school
3    Howard Jimenez    male  ...          5.56                             school

[4 rows x 10 columns]

In [160]:
dataset_multindex = pd.pivot_table(df_grades,
                                   index=['race/ethnicity', 'simple parental level of education'],
                                   #columns=['race/ethnicity', 'simple parental level of education','math score', 	'reading score', 	'writing score'],
                                   values=['math score', 	'reading score', 	'writing score'])
dataset_multindex.index.get_level_values(1)

Index(['college', 'postgraduate', 'school', 'college', 'postgraduate',
       'school', 'college', 'postgraduate', 'school', 'college',
       'postgraduate', 'school', 'college', 'postgraduate', 'school'],
      dtype='object', name='simple parental level of education')

**Resultado Esperado**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

## Visualizaciones [0.5 Puntos]

Ahora, implemente un gráfico de caja en donde se muestre el GPA con respecto al nivel educacional y que la variable de color sea la etnicidad y luego comente.

In [168]:
df2["GPA"] = pd.to_numeric(df2["GPA"])

In [182]:
df3 = pd.DataFrame(df2)
df3

math score  ...  percentage
race/ethnicity simple parental level of education              ...            
group A        college                                   4.73  ...       2.74%
               postgraduate                              4.69  ...       0.23%
               school                                    4.57  ...       5.83%
group B        college                                   5.07  ...       6.17%
               postgraduate                              4.91  ...       0.57%
               school                                    4.69  ...      12.23%
group C        college                                   5.02  ...      11.66%
               postgraduate                              4.92  ...       1.71%
               school                                    4.76  ...      17.71%
group D        college                                   5.11  ...        8.0%
               postgraduate                              5.22  ...       2.29%
               school                                    5.02  ...      17.03%
group E        college                                   5.54  ...       5.94%
               postgraduate                              5.54  ...       0.69%
               school                                    5.40  ...        7.2%

[15 rows x 6 columns]

In [183]:
import plotly.express as px

#fig = px.box(pd.DataFrame(df2), x=df2.index.get_level_values(1), y="GPA", color=df2.index.get_level_values(0))
#fig.update_traces(quartilemethod="exclusive") # or "inclusive", or "linear" by default
#fig.show()
fig = px.box(df3, x=df3.index.get_level_values(1), y="GPA", color=df2.index.get_level_values(0))
fig.show()

> 1. ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?
> 2. ¿Este gráfico permite hacer facilmente un análisis conjunto de estas dos variables de forma sencilla?

**Justifique:**

```
Escriba aqui su respuesta
```

# 1.3.5 Combinar Dataset [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias. 
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.
 
 
El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`). 


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**

- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

In [187]:
other_grades= pd.read_csv("other_grades.csv")
other_grades.head(3)

names  science score  history score
0      Pam Otoole             57            4.0
1     Diane Olson             48            4.8
2  Sarah Flannery             72            6.1

In [189]:
df_grades = pd.read_json("students_grades.json")
df_grades.head(3)

names  gender  ... reading score writing score
0    Rita Courtney  female  ...            46            46
1  Madeline Fuller  female  ...           142           142
2    Daniel Rester    male  ...           196           193

[3 rows x 9 columns]

In [211]:
#Unir df_grades con other_grades.csv usando outer join
df_outer= pd.merge(df_grades, other_grades, on=["names"], how='outer')
df_outer.head(3)

names  gender  ... science score history score
0    Rita Courtney  female  ...          25.0           3.8
1    Rita Courtney  female  ...          25.0           3.8
2  Madeline Fuller  female  ...           NaN           NaN

[3 rows x 11 columns]

In [218]:
df_outer.shape

(1400, 11)

In [212]:
df_outer.isnull().sum(). sort_values(ascending=False)

history score                  153
science score                  153
writing score                    0
reading score                    0
math score                       0
test preparation course          0
lunch                            0
parental level of education      0
race/ethnicity                   0
gender                           0
names                            0
dtype: int64

In [209]:
#Unir df_grades con other_grades.csv usando left join
df_left= pd.merge(df_grades, other_grades, on=["names"], how='left')
df_left.head(3)

names  gender  ... science score history score
0    Rita Courtney  female  ...          25.0           3.8
1  Madeline Fuller  female  ...           NaN           NaN
2    Daniel Rester    male  ...           NaN           NaN

[3 rows x 11 columns]

In [219]:
df_left.shape

(1400, 11)

In [206]:
df_left.isnull().sum(). sort_values(ascending=False)

history score                  153
science score                  153
writing score                    0
reading score                    0
math score                       0
test preparation course          0
lunch                            0
parental level of education      0
race/ethnicity                   0
gender                           0
names                            0
dtype: int64

In [210]:
#Unir df_grades con other_grades.csv usando right join
df_right= pd.merge(df_grades, other_grades, on=["names"], how='right')
df_right.head(3)

names  gender  ... science score history score
0      Pam Otoole  female  ...            57           4.0
1     Diane Olson  female  ...            48           4.8
2  Sarah Flannery  female  ...            72           6.1

[3 rows x 11 columns]

In [220]:
df_right.shape

(1247, 11)

In [208]:
df_right.isnull().sum(). sort_values(ascending=False)

history score                  0
science score                  0
writing score                  0
reading score                  0
math score                     0
test preparation course        0
lunch                          0
parental level of education    0
race/ethnicity                 0
gender                         0
names                          0
dtype: int64

In [217]:
#Unir df_grades con other_grades.csv usando inner
df_inner= pd.merge(df_grades, other_grades, on=["names"], how='inner')
df_inner.head(3)

names  gender  ... science score history score
0     Rita Courtney  female  ...            25           3.8
1     Rita Courtney  female  ...            25           3.8
2  Charles Linstrom    male  ...            56           5.2

[3 rows x 11 columns]

In [221]:
df_inner.shape

(1247, 11)

In [215]:
df_inner.isnull().sum(). sort_values(ascending=False)

history score                  0
science score                  0
writing score                  0
reading score                  0
math score                     0
test preparation course        0
lunch                          0
parental level of education    0
race/ethnicity                 0
gender                         0
names                          0
dtype: int64

> **Justificación:**


```
Escriba aquí su justificación
```

## 1.4 Más visualizaciones [Bonus - 0.5 puntos por cada Gráfico, máximo 1 punto]


**Desde aquí parte el bonus para el lab.**


<p align="center">
  <img src="https://media4.giphy.com/media/xT5LMWNOjGqJzUfyve/giphy.gif?cid=ecf05e47phkhvg6vadqvcoobvxwj9bzxxk38jjvkcc6eak6t&rid=giphy.gif&ct=g">
</p>




Genere dos visualizaciones extras que encuentre interesantes (y no triviales) con estos datos y explique sus resultados. Agrupe los atributos que estime convenientes.


**To-Do:**
- [ ] Generar dos nuevas visualizaciones con los datos y explicar que están representando.


> **NOTA: No utilice historia ni ciencias, son notas generadas aleatoriamente.**

# Conclusión
Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>

<br>
<center>
<img src="https://i.kym-cdn.com/photos/images/original/001/194/195/b18.png" width=100 height=50 />
</center>
<br>